import os
import pandas as pd

__author__ = "Kristina Sungeun Song"
__email__ = "kristina.song@usherbrooke.ca"

configfile: "config/config.yaml"


#############################
## Samples Information
#############################

SNO = config["snoRNA"]
SAMPLES = config["samples"]
READS = ["R1","R2"]

df = pd.read_csv('data/comparisons.tsv', sep=' ')
comparisons = [f'{cdn1}-{cdn2}' for cdn1, cdn2 in df.values]
comparisons_full = [f'results/deseq2/{cdn1}-{cdn2}.csv' for cdn1, cdn2 in df.values]

df_cond = pd.read_csv('data/design.tsv',delim_whitespace=True,usecols=['sample','condition'])
conditions = df_cond['condition'].unique().tolist()

#############################
## Load rules
#############################

include: "rules/functions.smk"
include: "rules/trim.smk"
include: "rules/fastqc.smk"
include: "rules/align.smk"
include: "rules/quantification.smk"
include: "rules/diffexp.smk"
include: "rules/metrics.smk"
include: "rules/splicing.smk"


#############################
## Target rules
#############################


#### COMPLETE
#fastqc = expand("results/fastqc/{step}/{sample}_{read}_fastqc.{ext}", sample=SAMPLES, read=READS, step=["pretrim","posttrim"], ext=["html","zip"])
#bedgraphs = expand("results/genomecov/{sample}.bedgraph", sample=SAMPLES)
#picard_files = expand("results/picard/{sample}.isize.{ext}", sample=SAMPLES, ext=["txt","pdf"])
#multiqc = "results/multiqc/multiqc_report.html"
#pca_plot = "results/pca/pca.svg"
#volcano_plot = expand("results/volcano_plot/{comp}.svg", comp=comparisons)
#sig_diffexp_genes = expand("results/volcano_plot/{comp}_sig_diffexp_genes_{change}.tsv", comp=comparisons, change=['up','down'])
#splicing_events = expand("results/voila/deltapsi/event_types/{comp}/filtered/bar_chart.svg", comp=comparisons)

#### NEED INTERNET CONNECTION
#GO_bar_charts = expand("results/GO/GO_{comp}_{change}regulated_genes.svg", comp=comparisons, change=['up','down'])

rule all:
    input:
        bedgraphs = expand("results/genomecov/{sample}.bedgraph", sample=SAMPLES)